src/bootstrap/bootstrap.py

```diff
@@ -629,6 +629,12 @@ def get_toml(self, key, section=None):
         'true'
         """
 
+        # Target is quoted in bootstrap configure.
+        # See https://github.com/rust-lang/rust/commit/1532fd8cd0db93f469e414f9da31ef083a44fcba
+        # However, we do not want to break target.x86_64-unknown-linux-gnu (without the quotes).
+        if section is not None and section.startswith("target.") and section != "target.x86_64-unknown-linux-gnu":
+            section = "target.'{}'".format(section[7:])
```

Comment:
This is not quite what I meant - you've hard-coded the specific case I called out earlier, but I want to fix it generally, not just for `target.x86_64-unknown-linux-gnu`. The proper fix would be some change to `section_match` to allow quotes around the section; maybe just `.replace("'", "")`? at this point I wonder if we should bring in a toml parser, but `configparse` doesn't work and anything else requires us to pull in a dependency :/

cc @Mark-Simulacrum, we're parsing .toml incorrectly in python because of the regex and I was wondering if you had suggestions for how to fix it.

Reply:
I guess we could vendor [tomllib](https://docs.python.org/3/library/tomllib.html) into bootstrap although I'm not super excited about that.

Reply:
I don't think we should vendor anything into our code.

Can you say more about "target is quoted" from the PR description? In particular, if the problem is that we don't support a particular syntax but there are workarounds, that doesn't really worry me.

In general, I think `--set` should be considered best-effort; users who want full fidelity can always directly append custom TOML to config.toml without too much trouble IMO. If we want to further support that I would be OK with a `--append file/path` option that inlines some other file(s) as well or similar, but I'm not sure it's really necessary.

Reply:
@Mark-Simulacrum some history: *we ourselves* are the ones adding quotes to the target name in `configure`: https://github.com/rust-lang/rust/blob/4151a5b96cb4707bb0b09037bacc58ade1940fa2/src/bootstrap/configure.py#L408
The reason for that was to allow targets that had a `.` in their name: https://github.com/rust-lang/rust/issues/97263
So we're generating configuration in `configure.py` that `bootstrap.py` can't read, which seems pretty bad to me.

The reason we didn't notice until now is that *only* bootstrap.py couldn't read it; the toml parsing is correct in rustbuild. I think @MarcusCalhoun-Lopez ran into issues because he's configuring the linker for a target and expecting that to be used for *rustbuild itself*, which was the case until we started quoting targets.

> In general, I think --set should be considered best-effort

I am not sure that's relevant here, since even modifying config.toml manually won't get us the right behavior; you'll still be unable to use quotes around a target name and have bootstrap.py parse it correctly.

Reply:
Yes, that does seem different (and talking about configure/--set doesn't seem entirely relevant, as you say). It sounds like we're always generating valid TOML, we just can't parse it later?

That said, it also seems like a valid workaround *is* generating TOML that bootstrap.py is capable of parsing, right? (Perhaps so long as your target's name doesn't contain `.`).

I think long-term https://docs.python.org/3/library/tomllib.html is the way to go; if we don't remove bootstrap.py's need for parsing complex TOML via the RIIR effort for rustbuild.

It will presumably be available on most distros in the next year or three...

Reply:
I'd also be okay (if this works) not quoting things that don't contain `.`. maybe that fixes our parsing too?

Reply:
> I'd also be okay (if this works) not quoting things that don't contain `.`. maybe that fixes our parsing too?

:+1:  that seems like a reasonable compromise in the meantime.

Reply:
I have updated the pull request that I believe addresses the concerns.